File: 11-flyweight/11-flyweight.py
class Letter:
    def __init__(self, char):
        self.char = char

    def render(self):
        print("This is a char:", self.char)

class CharFactory:
    def __init__(self):
        self.chars = {}

    def get_char(self, char):
        if char not in self.chars:
            self.chars[char] = Letter(char)
        return self.chars[char]

    def total_chars(self) -> int:
        return len(self.chars)

# Client code
c = CharFactory()
word = "mamma mia"
for l in word:
    if l.isalpha():
        char_obj = c.get_char(l)
        char_obj.render()

print("Total unique letters:", c.total_chars())

File: 11-flyweight/11-flyweight.js
class Character {
    constructor(character) {
        this.character = character;
    }

    render() {
        return this.character;
    }
}

class CharacterFactory {
    constructor() {
        this.chars = {};
    }

    getChar(char) {
        if (!this.chars[char]) {
            this.chars[char] = new Character(char);
        }
        return this.chars[char];
    }

    countChars() {
        return Object.keys(this.chars).length;
    }

    allChars() {
        return this.chars;
    }
}

// Client code
const cf = new CharacterFactory();
cf.getChar("a");
cf.getChar("b");
cf.getChar("a");

console.log(cf.allChars(), cf.countChars());
